File: internal/service/reembed.go
package service

import (
	"context"
	"fmt"
	"io"

	"uni-context/internal/port"
)

// ReembedService bulk-embeds items under the active model. The filter
// differs from BackfillService: BackfillService targets items where
// any_embedding=0 (first-time embed), while ReembedService targets items
// that lack a status='done' row for the active model (migration to a new
// active model after `embed switch`).
//
// Idempotent: re-runs skip items already done for the active model.
// Resumable: failed items get status='failed' rows and are picked up by
// the worker (which is model-agnostic).
type ReembedService struct {
	repo   port.ContextRepo
	embed  *EmbedService
	active port.ModelInfo // slug of the currently-wired embedder
	// log receives per-100-item progress lines. Injected via constructor
	// so tests can assert on progress and the service has no direct
	// os.Stderr coupling.
	log io.Writer
}

// NewReembedService wires the ContextRepo (lists candidate items) +
// EmbedService (embeds each) + the active model identifier (filter key)
// + a logger for progress reporting.
func NewReembedService(repo port.ContextRepo, embed *EmbedService, active port.ModelInfo, log io.Writer) *ReembedService {
	return &ReembedService{repo: repo, embed: embed, active: active, log: log}
}

// ReembedFailure records a per-item embed error. Aggregated in
// ReembedService.Run's report so the CLI can surface them.
type ReembedFailure struct {
	ItemID string
	Error  string
}

// ReembedReport summarizes one Run invocation. Scanned = candidates
// found (no done row for active model); Embedded = successful embeds;
// Failed = per-item failures.
type ReembedReport struct {
	Scanned  int
	Embedded int
	Failed   int
	Failures []ReembedFailure
}

// Run iterates items lacking a status='done' row for the active model
// and embeds each. For each item:
//   - dryRun=true: increment Scanned only.
//   - dryRun=false: call EmbedService.Embed; on failure record a
//     ReembedFailure and continue; on success increment Embedded.
//
// limit<=0 means no limit. Progress is logged to stderr every 100 items.
//
// Run does NOT return an error on per-item embed failures; the only error
// it returns is from the initial List call or ctx cancellation.
func (s *ReembedService) Run(ctx context.Context, limit int, dryRun bool) (ReembedReport, error) {
	var report ReembedReport

	items, _, err := s.repo.List(ctx, port.ItemFilter{
		NotDoneForModel: s.active.Slug,
		Limit:           limit,
	})
	if err != nil {
		return report, fmt.Errorf("list items pending for model %s: %w", s.active.Slug, err)
	}

	for i, item := range items {
		select {
		case <-ctx.Done():
			return report, ctx.Err()
		default:
		}

		report.Scanned++
		if dryRun {
			continue
		}

		if err := s.embed.Embed(ctx, item.ID, item.Title, item.Content); err != nil {
			report.Failed++
			report.Failures = append(report.Failures, ReembedFailure{
				ItemID: item.ID,
				Error:  err.Error(),
			})
			continue
		}
		report.Embedded++

		if (i+1)%100 == 0 {
			fmt.Fprintf(s.log, "reembed: %d items processed\n", i+1)
		}
	}
	return report, nil
}

File: python/src/unictx/embed/reembed.py
"""ReembedService — bulk re-embed items under the active model.

Behavior-port of Go's ``internal/service/reembed.go``. Differs from
BackfillService in filter: BackfillService targets items where
``any_embedding=0`` (first-time embed); ReembedService targets items
that lack a ``status='done'`` row for the active model (migration to
a new model after ``embed switch``).

Plan §Python Conventions adaptations:
  - ctx dropped (Python is sync). Cancellation via ``threading.Event``.
  - Go's ``Run(ctx, limit, dryRun) (ReembedReport, error)`` →
    ``run(limit, dry_run) -> ReembedReport``. List error raises.
  - ``log`` defaults to ``sys.stderr``; tests pass ``StringIO``.

Idempotent (Go reembed.go:17-19 preserved): re-runs skip items already
done for the active model. Resumable: failed items get status='failed'
rows and are picked up by WorkerService (which is model-agnostic).
"""

from __future__ import annotations

import contextlib
import sys
import threading
from dataclasses import dataclass, field
from typing import IO

from unictx.embed.embedder import ModelInfo
from unictx.embed.service import EmbedService
from unictx.items.repo import ContextRepo, ItemFilter

__all__ = ["ReembedService", "ReembedReport", "ReembedFailure"]


@dataclass(slots=True)
class ReembedFailure:
    """One per-item embed error during a reembed run."""

    item_id: str
    error: str


@dataclass(slots=True)
class ReembedReport:
    """Summarizes one ``ReembedService.run`` invocation.

    Mirrors Go's ReembedReport (reembed.go:47-52). ``scanned`` counts
    candidates found (no done row for active model); ``embedded``
    counts successful embeds; ``failed`` counts per-item failures.
    """

    scanned: int = 0
    embedded: int = 0
    failed: int = 0
    failures: list[ReembedFailure] = field(default_factory=list)


class ReembedService:
    """Bulk re-embeds items lacking a status='done' row for the active model.

    Construction is cheap (no I/O). The ``active`` ModelInfo identifies
    the currently-wired embedder — its slug becomes the
    ``ItemFilter.not_done_for_model`` filter key.

    Idempotent + resumable. Re-runs after a partial success skip items
    already done for the active model. Failures get status='failed'
    rows and are picked up by the WorkerService.
    """

    def __init__(
        self,
        repo: ContextRepo,
        embed_svc: EmbedService,
        active: ModelInfo,
        log: IO[str] | None = None,
    ) -> None:
        self._repo = repo
        self._embed = embed_svc
        self._active = active
        self._log: IO[str] = log if log is not None else sys.stderr

    def run(
        self,
        limit: int = 0,
        dry_run: bool = False,
        stop_event: threading.Event | None = None,
    ) -> ReembedReport:
        """Iterate items lacking a status='done' row for the active model.

        Per item:
          - ``dry_run=True``: increment ``scanned`` only.
          - ``dry_run=False``: call EmbedService.embed_item; on failure
            record a ``ReembedFailure`` and continue; on success
            increment ``embedded``.

        ``limit<=0`` means no limit.

        Progress is logged every 100 items. Mirrors Go reembed.go:97-99.
        """
        report = ReembedReport()

        items, _next_cursor = self._repo.list(
            ItemFilter(not_done_for_model=self._active.slug, limit=limit)
        )

        for i, item in enumerate(items):
            if stop_event is not None and stop_event.is_set():
                return report

            report.scanned += 1
            if dry_run:
                continue

            try:
                self._embed.embed_item(item.id, item.title, item.content)
            except Exception as exc:
                report.failed += 1
                report.failures.append(
                    ReembedFailure(item_id=item.id, error=str(exc))
                )
                continue
            report.embedded += 1

            if (i + 1) % 100 == 0:
                self._warn(f"reembed: {i + 1} items processed\n")

        return report

    # ---- internals ---------------------------------------------------

    def _warn(self, msg: str) -> None:
        """Write a warning/progress line to the injected log. Best-effort."""
        with contextlib.suppress(Exception):
            self._log.write(msg)
